nixos/modules/services/torrent/btpd.nix

```diff
@@ -0,0 +1,191 @@
+        '';
+      };
+
+      bandwidthLimitIn = mkOption {
+        type = types.ints.unsigned;
+        default = 0;
+        description = ''
+          Limit incoming BitTorrent traffic to n kB/s. 0 is unlimited.
+        '';
+      };
+
+      bandwidthLimitOut = mkOption {
+        type = types.ints.unsigned;
+        default = 0;
+        description = ''
+          Limit outgoing BitTorrent traffic to n kB/s. 0 is unlimited.
+        '';
+      };
+
+      emptyStart = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Start btpd without any active torrents.
+        '';
+      };
+
+      ipcPermission = mkOption {
+        type = types.ints.unsigned;
+        default = 600;
+        example = 666;
+        description = ''
+          Permission mode of the btpd socket.
+        '';
+      };
+
+      maxPeers = mkOption {
+        type = types.ints.unsigned;
+        default = 0;
+        description = ''
+          Limit the amount of peers to n.
+        '';
+      };
+
+      maxUploads = mkOption {
+        type = types.int;
+        default = -2;
+        description = ''
+          Controls the number of simultaneous uploads.
+          The possible values are:
+          <itemizedlist>
+            <listitem><para>
+              <literal>n &lt; -1</literal>: Choose n >= 2 based on bandwidthLimitOut (default).
+            </para></listitem>
+            <listitem><para>
+              <literal>n = -1</literal>: Upload to every interested peer.
+            </para></listitem>
+            <listitem><para>
+              <literal>n = 0</literal>: Don't upload to anyone.
+            </para></listitem>
+            <listitem><para>
+              <literal>n &gt; 0</literal>: Upload to at most n peers simultaneously.
+            </para></listitem>
+          </itemizedlist>
+        '';
+      };
+
+      prealloc = mkOption {
+        type = types.ints.unsigned;
+        default = 2048;
+        description = ''
+          Preallocate disk space in chunks of n kB.
+          Note that n will be rounded up to the closest multiple of the
+          torrent piece size. If n is zero no preallocation will be done.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.btpd ];
+
+    users.users.btpd = mkIf (cfg.user == "btpd") { group = mkDefault "btpd"; };
+    users.groups.btpd = mkIf (cfg.group == "btpd") { };
+
+    systemd.services.btpd = {
+      after = [ "network.target" ];
+      description = "BitTorrent Protocol Daemon";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        StateDirectory = mkIf (cfg.dataDir == "/var/lib/btpd") [ "btpd" ];
+        ExecStart = ''
+          ${pkgs.btpd}/bin/btpd \
+              --no-daemon \
+              -d ${escapeShellArg cfg.dataDir} \
```

Comment:
You should probably ensure this directory exists. As suggested `systemd.tmpfiles.rules` would be the appropriate way to do so.